keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,843 @@
+aware of the pods and containers running on the machine and systemd will simply
+kill them as regular linux processes.
+
+### Background on Inhibitors
+
+`systemd-logind` provides the ability for applications to delay shutdown and
+perform some series of actions before the shutdown completes through a
+mechanism called ["Inhibitor
+Locks".](https://www.freedesktop.org/wiki/Software/systemd/inhibit/)
+Applications can request to delay shutdown by taking an inhibitor lock by
+sending messages to logind over dbus. Applications can request up to
+`InhibitDelayMaxSec` (a setting configured in logind.conf`) for delay based
+locks, which allow applications to receive sleep and shutdown events, and block
+the shutdown from proceeding by `InhibitDelayMaxSec` period to execute some
+critical work prior to shutdown/sleep.
+
+We believe that making use of systemd is a reasonable approach considering
+almost all new popular linux distros are systemd based (RHEL, Google COS,
+Ubuntu, CentOS, Debian, Fedora, Flatcar Linux, see widespread
+[adoption](https://en.wikipedia.org/wiki/Systemd#Adoption)) and [systemd
+183](https://lwn.net/Articles/499480/) (released in 2012) features support for
+inhibitors.
+
+Thanks to @giuseppe for helping with getting systemd inhibitors working!
+
+### Implementation
+
+Introduce a new Kubelet Config setting, `kubeletConfig.ShutdownGracePeriod`,
+defaulting to 0 seconds. Upon kubelet startup,
+
+*   if the setting is greater than 0 seconds
+    *   kubelet will check with dbus current `InhibitDelayMaxSec` to check if
+        `kubeletConfig.ShutdownGracePeriod &lt;=` `InhibitDelayMaxSec`.
+*   if `kubeletConfig.ShutdownGracePeriod` > `InhibitDelayMaxSec`
+    *   Kubelet will attempt to update the InhibitDelayMaxSec setting, by
+        writing a config file to `/etc/systemd/logind.conf.d/kubelet.conf`, and
+        sending a SIGHUP to logind to update the config setting to ensure that
+        the ShutdownGracePeriod from kubelet config is equal to
+        `InhibitDelayMaxSec`.
+
+After updating the `InhibitDelayMaxSec` on the node if needed, Kubelet will
+query the dbus for the final value of `InhibitDelayMaxSec` set on the node and
+treat min(`InhibitDelayMaxSec`, `kubeletConfig.ShutdownGracePeriod`) as the
+allocatable  shutdown grace period, which will be referred to in this KEP as
+`ShutdownGracePeriod`.
+
+Kubelet will register with dbus as a delay systemd inhibitor lock for the
+`ShutdownGracePeriod` for the shutdown event. Kubelet will also register a
+`PrepareForShutdown` signal which will be emitted prior to the shutdown. Upon
+receiving the signal, Kubelet will have additional `ShutdownGracePeriod` time
+before the actual node will initiate the shutdown.
+
+**Handling the shutdown**
+
+Upon a shutdown occurring, Kubelet will gracefully terminate all the pods
+running on the node and taint itself with a new taint,
+`node.kubernetes.io/unschedulable` to avoid new pods being scheduled.
```

Comment:
will this taint removed when machine is back up after the restart?

Reply:
That's the goal, ideally kubelet can update the status of node and add this taint when the shutdown starts. If the node is rebooted or another node comes up with the same name, ideally the node will update the status without this taint. Still a bit TBD, figuring out how this will look like implementation wise.